#163388139 users should be able to rate articles

asirvex opened the pull request:
**What does this PR do?**
creates a feature for the users to rate articles

**Description of tasks to be completed**
- Write tests for the rating functionality
- Create the views, models and serializers for the feature
- refactor some of the basetest methods

**Any background context you want to provide?**

- An article can be rated by an authenticated users
- Any user can see the articles ratings
- A user should not be able to rate their own artilces

**How should this be manually tested?**

1. Run the server:
 `python manage.py runserver`

2. From postman:
   
   1. Create an article by sending a `post` request to the '/api/v1/articles/' endpoint with the data body
   
   ```
   {
       "article": {
       "title": "Test Article",
       "description": "Test description for the article",
       "body": "Test body for the article"
       }
   }
   ```
   2. Rate the article using the endpoint '/api/v1/rate/test-article-slug/' with the data body with
     with another registered user login credentials
   ```
   {
      "user_rating": "<decimal number between 1-5>"
   }
   ```
   3. Get an articles ratings using the endpoint '/api/v1/rate/test-article-slug/'

[#163388139](https://www.pivotaltracker.com/story/show/163388139)

asirvex committed:
Feature(authors haven): implement rating system on the articles
- create models for the ratings
- create the rating views
- create serializers for the ratings
[Start #163388139]

asirvex committed:
feature(authors haven): create the article rating feature
- create rating models
- create the rating serializers
- create the rating views

asirvex committed:
feature(authors haven): refactor basetest methods
- change authenticate_user to take in a varible

asirvex committed:
feature(authors haven):

asirvex committed:
feature(authors haven): article ratings
- finish the ratings feature
- create passing tests for the feature
- refactor reusable test methods

[Finishes #163388139]

PaulinhoNdegwa commented on authors/apps/articles/serializers.py:
```diff
@@ -27,7 +28,45 @@ class ArticleSerializer(serializers.ModelSerializer):
 
+    def retrieve_author(self, obj):
+        """retrieve the author profile"""
+        pass
```
@asirvex To improve your code, you can remove this method since you are not using it

asirvex replied:
Thanks for the feedback, I'll remove it

asirvex committed:
feature(authors haven) implement rating for the article
- Remove an unused function in the article class

[Finishes #163388139]

sirghiny commented on authors/apps/articles/models.py:
```diff
@@ -41,3 +39,21 @@ def save(self, *args, **kwargs):
+        related_name="rated_article",
+        on_delete=models.CASCADE
+    )
+    user_rating = models.FloatField()
```
Make thus un-nullable.

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -84,4 +99,89 @@ def destroy(self, request, pk=None):
+            raise ValidationError(
+                detail={'message': 'Article not found'}
+            )
+        if request.user.is_authenticated is False:
```
This should be done by the `permission_class`

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -84,4 +99,89 @@ def destroy(self, request, pk=None):
+            if avg["user_rating__avg"] is None:
+                average_rating = 0
+
+            if request.user.is_authenticated is False:
```
Should be handled by the `permissions_class`

asirvex replied:
this route allows both authenticated and unauthenticated users to access it but returns different results for each scenario

sirghiny replied:
Why are we allowing unauthenticated users to access anything?

asirvex replied:
Unauthenticated users can read an article and also see an articles ratings

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -84,4 +99,89 @@ def destroy(self, request, pk=None):
+    serializer_class = RatingSerializer
+    permission_class = (IsAuthenticatedOrReadOnly,)
+
+    def get_article(self, slug):
```
Isn't this handled by the helper above? (line 15)

dev-jey commented on authors/apps/articles/tests/test_ratings.py:
```diff
@@ -0,0 +1,97 @@
+from rest_framework import status
+
+
+class TestArticleRatings(BaseTest):
```
Add a doc string to the class

asirvex committed:
feature(authors haven): a user can rate an article
- refactor the code to use permission classes where necessary
- remove the rendunt get_article function
[Finishes #163388139]

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -105,26 +105,18 @@ def destroy(self, request, pk=None):
-        article = Article.objects.all().filter(slug=slug).first()
-        return article
+    permission_classes = (IsAuthenticatedOrReadOnly,)
+    # renderer_classes = (RatingJsonRenderer,)
```
Remove commented out code.

sirghiny commented on authors/apps/articles/serializers.py:
```diff
@@ -31,3 +32,37 @@ class Meta:
+
+    def get_article(self, obj):
+        """Returns an article that matches the slug"""
+        return obj.article.slug
```
Will this return the Article or the article's slug?

asirvex replied:
It will return the article's slug. Let me change the variable name and docstring

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -84,4 +99,81 @@ def destroy(self, request, pk=None):
+        rating = request.data
+        article = get_article(slug)
+
+        if not article:
```
This won't happen since in the absence of an article, an exception will be thrown (line 20)

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -1,13 +1,28 @@
+            {"error": "Article not found"},
+            status=status.HTTP_404_NOT_FOUND
+        )
+    return article
```
This is better right after the attempt to get an article in line 18. Always complete a try block before moving to the Except.

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -84,4 +99,81 @@ def destroy(self, request, pk=None):
+    def get(self, request, slug):
+        """Get request for an article ratings."""
+        rating = None
+        try:
```
Can't you use the method defined in line 15 as it was used in line 114?

asirvex replied:
When I use it I get the error I informed you about

asirvex committed:
feature(authors haven): users can rate an article
- remove commented code
- refactor the code to remove redundant functions
- fix crashing of the application when a user who has not rated the article gets the ratings

[Finishes #163388139]
